index.js

```diff
@@ -1,376 +1,340 @@
+        indegree: indegree,
+        outdegree: outdegree,
+        depthFirstSearch: depthFirstSearch,
+        lowestCommonAncestors: lowestCommonAncestors,
+        topologicalSort: topologicalSort,
+        shortestPath: shortestPath,
+        serialize: serialize,
+        deserialize: deserialize
+    };
+    // The adjacency list of the graph.
+    // Keys are node ids.
+    // Values are adjacent node id arrays.
+    var edges = {};
+    // The weights of edges.
+    // Keys are string encodings of edges.
+    // Values are weights (numbers).
+    var edgeWeights = {};
+    // If a serialized graph was passed into the constructor, deserialize it.
+    if (serialized) {
+        deserialize(serialized);
     }
-
-    return graph;
-  }
-
-  // Removes the edge from node u to node v.
-  // Does not remove the nodes.
-  // Does nothing if the edge does not exist.
-  function removeEdge(u, v){
-    if(edges[u]){
-      edges[u] = adjacent(u).filter(function (_v){
-        return _v !== v;
-      });
+    // Adds a node to the graph.
+    // If node was already added, this function does nothing.
+    // If node was not already added, this function sets up an empty adjacency list.
+    function addNode(node) {
+        edges[node] = adjacent(node);
+        return graph;
     }
-    return graph;
-  }
-
-  // Computes the indegree for the given node.
-  // Not very efficient, costs O(E) where E = number of edges.
-  function indegree(node){
-    var degree = 0;
-    function check(v){
-      if(v === node){
-        degree++;
-      }
+    // Removes a node from the graph.
+    // Also removes incoming and outgoing edges.
+    function removeNode(node) {
+        // Remove incoming edges.
+        Object.keys(edges).forEach(function (u) {
+            edges[u].forEach(function (v) {
+                if (v === node) {
+                    removeEdge(u, v);
+                }
+            });
+        });
+        // Remove outgoing edges (and signal that the node no longer exists).
+        delete edges[node];
+        return graph;
     }
-    Object.keys(edges).forEach(function (u){
-      edges[u].forEach(check);
-    });
-    return degree;
-  }
-
-  // Computes the outdegree for the given node.
-  function outdegree(node){
-    return node in edges ? edges[node].length : 0;
-  }
-
-  // Depth First Search algorithm, inspired by
-  // Cormen et al. "Introduction to Algorithms" 3rd Ed. p. 604
-  // This variant includes an additional option
-  // `includeSourceNodes` to specify whether to include or
-  // exclude the source nodes from the result (true by default).
-  // If `sourceNodes` is not specified, all nodes in the graph
-  // are used as source nodes.
-  function depthFirstSearch(sourceNodes, includeSourceNodes){
-
-    if(!sourceNodes){
-      sourceNodes = nodes();
+    // Gets the list of nodes that have been added to the graph.
+    function nodes() {
+        // TODO: Better implementation with set data structure
```

Comment:
Reasonable I guess, but generally I prefer to keep `TODO`s as issues rather than comments.

Created a new issue for this: [Use ES6 Map and Set where appropriate](https://github.com/datavis-tech/graph-data-structure/issues/27).